cocotb/scheduler.py

```diff
@@ -752,20 +757,20 @@ def schedule(self, coroutine, trigger=None):
         # chaining
         if coro_completed:
             self.unschedule(coroutine)
-            return
 
         # Don't handle the result if we're shutting down
         if self._terminate:
             return
 
-        try:
-            result = self._trigger_from_any(result)
-        except TypeError as exc:
-            # restart this coroutine with an exception object telling it that
-            # it wasn't allowed to yield that
-            result = NullTrigger(outcome=outcomes.Error(exc))
+        if not coro_completed:
+            try:
+                result = self._trigger_from_any(result)
+            except TypeError as exc:
+                # restart this coroutine with an exception object telling it that
+                # it wasn't allowed to yield that
+                result = NullTrigger(outcome=outcomes.Error(exc))
 
-        self._coroutine_yielded(coroutine, result)
+            self._coroutine_yielded(coroutine, result)
```

Comment:
In my opinion this new conditional is a bit messy, but so is all the handling of external / function - so fine as is for now. Cleaning this up is on my todo list, and it will be a lot easier with this race condition fixed.